mio/src/lib.rs

```diff
@@ -316,21 +394,66 @@ impl Loop {
             Err(e) => futures::failed(e).boxed(),
         }
     }
+*/
 
     pub fn tcp_listen(&mut self, addr: &SocketAddr) -> io::Result<TcpListener> {
         let tcp = try!(mio::tcp::TcpListener::bind(addr));
+
+        // dummy registration, so that we can always re-register in the future
         try!(self.io.register(&tcp,
                               mio::Token(0),
                               mio::EventSet::none(),
                               mio::PollOpt::empty()));
 
         Ok(TcpListener {
             tcp: Arc::new(tcp),
-            tx: self.tx.clone(),
+            loop_handle: self.handle(),
         })
     }
 }
 
+impl TokenCounter {
+    pub fn new() -> TokenCounter {
+        TokenCounter { counter: Arc::new(AtomicUsize::new(1)) }
+    }
+
+    pub fn next_token(&self) -> usize {
+        // TODO: handle rollover robustly...
+        // the 0 token is reserved
+        let mut next = 0;
+        while next == 0 {
+            next = self.counter.fetch_add(1, Ordering::Relaxed);
+        }
+        next
+    }
+}
+
+// TODO: use TLS to avoid sending messages
+impl LoopHandle {
+    fn register(&self, source: Source) {
+        self.tx.send(Message::Register(source))
+            .map_err(|_| ())
+            .expect("failed to send register message") // todo: handle failure
```

Comment:
This probably wants to get punted upwards immediately, and that way the futures themselves can squirrel away this error, immediately call the wake callback, and then return the error on the next call to `poll`